### PR TITLE
feat(comp): environment variable completion with prefix filtering (#225)

### PR DESCRIPTION
**Pull request by WayDownWatson**
Implements two environment variable completers:
- CompVars: completes with $ prefix (e.g., $HOSTNAME)
- CompNames: completes without prefix (e.g., HOSTNAME)

Configuration options:
- Prefix filtering: restrict completions to vars with specific prefix Example: when "APP_" only matches APP_* variables
- Case sensitivity: optional case-insensitive matching Example: "path" matches PATH, Path, path, etc.

Compatibility Issues:

After implementation, I found that both bash and zsh have case-insensitive configuration. I left the implementation in place in order to discuss where you wanted to draw the line on shell configuration vs bonzai/completer configuration.

There are inconsistent behaviors with zsh and bash for this completer. I have not been able to find configuration that is able to resolve the inconsistencies. The CompVars completer appears to be useless in zsh (unless someone else knows of configuration that handles this) so there is a question in my mind it should be included.

- When a prefix is configured (`XDG_`)
   - both shells complete `cmd <TAB>` to `XDG_`
   - both shells complete `cmd X<TAB>` to `XDG_` where continued completion behaves correctly.
   - bash will complete `cmd C<Tab>` as `cmd XDG_CURRENT_DESKTOP`
   - zsh will not complete `cmd C<Tab>`, it will however complete `cmd XDG_C` as `cmd XDG_CURRENT_DESKTOP`
   It appears zsh will only complete items offered in the completion list if they have a prefix of the typed characters.
- When the `CompVars` completer is used
  - if a prefix is configured (`XDG_`), zsh will complete `cmd <Tab>` to `cmd \$XDG`, which blocks continued completion.
  - if a prefix is not configured, zsh will not expand `cmd HO<Tab>`, the only way to expand an env var is with `cmd $HO<Tab>`, where builtin completion takes over
  - bash will complete the above examples as `cmd $XDG` and `cmd HO<Tab>` as `cmd $HOME`

### CODE REVIEW

**Comment by rwxrob** [comment]
I absolutely love this, but there are a few things we'll need to change:

- Perhaps drop `Comp` from the name (`env.XDGCompNames` -> `env.XDGNames`)
- Replace run-time constructors with composable notation (`var XDGCompNames = env.New*` -> `Comp: env.XDGNames{CaseSensitive: false}`)

By using composition we can also allow a bunch of preconfigured ones using `comp.Combine` and just assignment:

```
// in env
var XDG = env.Prefix{`XDG_`}
```

FYI, this can be done with `comp.Combine` as well using a filter:

```
var XDG = comp.Combine{env.Values,filt.Prefix{`XDG`}}
```

Seeing that makes be realize we probably want `comp.Env` in the main/core completers instead of in its own completer package. I imagine a lot of people will want that. Then this all gets really simple and clean in code:

```go
var Cmd = *bonzai.Cmd{
  Name: `foo`,
  Comp: comp.Env{`XDG`},
  Do: bonzai.Nothing,
}
```

And maybe even support for regx matching:

```go
var Cmd = *bonzai.Cmd{
  Name: `foo`,
  Comp: comp.Env{Expr:`^XDG`},
  Do: bonzai.Nothing,
}
```

I absolutely love that these completers can actually double as legit functions for looking up data anywhere else in code:

```go
func main(){
  // all XDG variables found in current environment
  names := comp.Env{`XDG`}
  fmt.Println(names)
}
```

In other words, there should be no need for runtime stuff other than simple assignment of new instances using composition (even though technically that is still runtime, but without the function call indirection).

**Comment by WayDownWatson** [comment]
@rwxrob Is this what you had in mind?

I left the code in completers/env for now. I can move it to comp.Env if I'm on the right track.

I also did not yet define any preconfigured, but the example & test show how Combine can now do what CompVars was doing previously. I do assume you meant tr.Prefix rather than filter.Prefix above?

**Comment by Chaitanyabsprip** [comment]
I wonder what would be the usecase for completing environment variables after the root command. Could either of you please provide an example usecase?

**Comment by rwxrob** [comment]
This looks good. Cmd.Env is going away, but let's still put it in `comp.Env`.
